code/GridFieldOrderableRows.php

```diff
@@ -450,10 +485,10 @@ protected function reorderItems($list, array $values, array $order) {
 			// *_versions table is updated. This ensures re-ordering works
 			// similar to the SiteTree where you change the position, and then
 			// you go into the record and publish it.
-			foreach(array_values($order) as $pos => $id) {
```

Comment:
using array_values here to force contiguous sort values was making it impossible to insert new records in the same postback as existin records. Note that here $sortedIDs will have gaps for record sort values handled by GridFieldAddNewInlineButton (or others).